src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -600,6 +602,80 @@ class ChooseStringGenerator : public Generator<std::string> {
     bool _deterministic;
 };
 
+/**
+ * FrequencyMaps
+ * A list of {string, count} pairs.
+ * An item will be pulled from a random bucket and its count will be decremented. If all buckets are
+ * empty, the generator throws an error.
+ */
+class FrequencyMapGenerator : public Generator<std::string> {
+public:
+    FrequencyMapGenerator(const Node& node, GeneratorArgs generatorArgs) : _rng{generatorArgs.rng} {
+        if (!node["from"].isMap()) {
+            std::stringstream msg;
+            msg << "Malformed node for 'TakeRandomStringFromFrequencyMap' from a map " << node;
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+        }
+
+        for (const auto&& [k, v] : node["from"]) {
+            _map.push_back(v.key(), static_cast<size_t>(v.maybe<std::int64_t>().value()));
+        }
+    }
+
+    std::string evaluate() override {
+        if (_map.size() == 1) {
+            return _map.take(0);
+        }
+
+        // Pick a random number between 0 and _map.size() inclusive
+        auto distribution = boost::random::uniform_int_distribution<size_t>(0, _map.size() - 1);
+        auto value = distribution(_rng);
+        return _map.take(value);
```

Comment:
Nit, but consider putting this logic into the `_map` - just pass in an RNG to `take`. This knows a lot about `_map`'s internals.

This only feels like an important consideration because there's a separate header for it.

Again, just something to consider.